tools/graphemebreak_test.c

```diff
@@ -0,0 +1,123 @@
+/* Simple program that read the grapheme break test database of the unicode standard
+ * and checks the output of libunibreak against the desired values of the test file
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <graphemebreak.h>
+
+static int hex[256] = {
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  0,  0,  0,  0,  0,  0,
+   0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+   0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+};
+
+int
+main(int argc, char *argv[])
+{
+  FILE *fp = fopen("GraphemeBreakTest.txt", "r");
+  int i;
+  char line[500];
+  char breaks_desired[500];
+  char breaks_actual[500];
+  utf32_t txt[500];
+  int linenumber = 1;
+
+  if (fp == 0)
+  {
+    fprintf(stderr, "Error opening file");
+    return 1;
+  }
+
+  while(fgets(line, sizeof(line), fp))//no line will be longer than 100
+  {
+    // remove everything from the first # on
+
+    int l = strlen(line);
+    int p = 0;
+
+    for (i = 0; i < l; i++)
+    {
+      if (line[i] == '#') line[i] = 0;
+    }
+
+    l = strlen(line);
+
+    // skip empty lines
+    if (l >= 5)
+    {
+      for (i = 0; i < 500; i++)
+      {
+        txt[i] = 0;
+      }
+
+      for (i = 3; i < l; i++)
```

Comment:
Why i=3? To skip the first break opportunity? I'd actually include it, and just have the array we fill off by one with the first one set to "break" so this code is less fragile, and we easily find out if they ever change the requirement that there's a break at the start.